src/main/java/com/hivemq/persistence/RetainedMessage.java

```diff
@@ -85,6 +114,52 @@ public RetainedMessage(@NotNull final PUBLISH publish, @Nullable final Long payl
         this.timestamp = publish.getTimestamp();
     }
 
+    public RetainedMessage copyWithoutPayload() {
+        return new RetainedMessage(
+                null,
+                qos,
+                payloadId,
+                messageExpiryInterval,
+                userProperties,
+                responseTopic,
+                contentType,
+                correlationData,
+                payloadFormatIndicator,
+                timestamp);
+    }
+
+    public int getEstimatedSizeInMemory() {
+
+        if (sizeInMemory != SIZE_NOT_CALCULATED) {
+            return sizeInMemory;
+        }
+        int size = 0;
+        if (message != null) {
+            size += message.length;
```

Comment:
I would just ignore the payload size. As the data is never at rest with the payload attached.